lib/dry/system/components/bootable.rb

```diff
@@ -1,37 +1,111 @@
+require 'dry/system/lifecycle'
+require 'dry/system/components/config'
+
 module Dry
   module System
     module Components
       class Bootable
+        DEFAULT_FINALIZE = proc {}
+
         attr_reader :identifier
 
-        attr_reader :fn
+        attr_reader :finalize
 
         attr_reader :options
 
-        def initialize(identifier, fn, options = {})
+        attr_reader :triggers
+
+        attr_reader :config
+
+        attr_reader :namespace
+
+        def initialize(identifier, options = {}, &block)
           @identifier = identifier
-          @fn = fn
-          @options = options
+          @triggers = { before: Hash.new { |h, k| h[k] = [] }, after: Hash.new { |h, k| h[k] = [] } }
+          @options = block ? options.merge(block: block) : options
+          @namespace = options[:namespace]
+          finalize = options[:finalize] || DEFAULT_FINALIZE
+          instance_exec(&finalize)
+        end
+
+        def block
+          options.fetch(:block)
+        end
+
+        def lifecycle
+          @lifecycle ||= Lifecycle.new(lf_container, component: self, &block)
+        end
+
+        def lf_container
+          container = Dry::Container.new
+
+          case namespace
+          when String, Symbol
+            container.namespace(namespace) { |c| return c }
+          when true
+            container.namespace(identifier) { |c| return c }
+          when nil
+            container
+          else
+            raise RuntimeError, "+namespace+ boot option must be true, string or symbol #{namespace.inspect} given."
+          end
+        end
+
+        def init
+          lifecycle.(:init)
+          trigger(:after, :init)
+          self
         end
 
-        def to_proc
-          fn
+        def start
+          lifecycle.(:start)
+          trigger(:after, :start)
+          self
+        end
+
+        def stop
+          lifecycle.(:stop)
+          self
+        end
+
+        def finalize
+          lifecycle.container.each do |key, item|
+            container.register(key, item) unless container.key?(key)
```

Comment:
I wonder if we should print a warning if there's a conflicting key? Or otherwise we could risk having components that are not fully registered in the main app container?

Reply:
@timriley good point, I now wonder if this should raise an exception, it probably should

Reply:
@timriley just realized we do want this check, it is possible that a component was already registered and it's actually fine. See `external_components_spec.rb` where we boot `:logger` from external provider 2 times.